pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -534,6 +558,52 @@ impl ImageLayerInner {
         Ok(planner.finish())
     }
 
+    /// Given a key range and filter function, generate a read plan and execute it, then for
+    /// each value we read write it onwards into the writer.
+    pub(super) async fn do_reads_and_write<F>(
```

Comment:
For the delta copy lsn prefix I did not reuse the existing vectored reads system completely because in the case that majority of the file needs to be read, we are unnecessarily building up those planned reads where as we could just process them as soon as they are built.

I guess you were thinking that you do the size estimate already, so that is not needed here? But then it would for me suggest that we shouldn't make this generic over the filter function.

Reply:
This isn't a generic filter function any more.

The plan for a rewrite only becomes large if someone uses a pathologically small stripe size (and users don't have control over that).